src/libclient/measurement/extupnp/extupnp.cpp

```diff
@@ -0,0 +1,295 @@
+#include "extupnp.h"
+
+#include "../../log/logger.h"
+#include <QMetaEnum>
+#include "../../types.h"
+
+LOGGER(ExtUPnP);
+
+#include <miniupnpc/miniupnpc.h>
+#include <miniupnpc/upnpcommands.h>
+#include <miniupnpc/miniwget.h>
+
+#ifndef FALSE
+#define FALSE 0
+#define TRUE 1
+#endif
+
+ExtUPnP::ExtUPnP(QObject *parent)
+: Measurement(parent)
+{
+}
+
+ExtUPnP::~ExtUPnP()
+{
+}
+
+Measurement::Status ExtUPnP::status() const
+{
+    return Unknown;
+}
+
+bool ExtUPnP::prepare(NetworkManager *networkManager, const MeasurementDefinitionPtr &measurementDefinition)
+{
+    Q_UNUSED(networkManager);
+    Q_UNUSED(measurementDefinition);
+    return true;
+}
+
+QStringList GetValsFromNameValueList(struct NameValueParserData *pdata,
+                                       const char *Name)
+{
+    QStringList ret;
+    struct NameValue *nv;
+
+    for (nv = pdata->l_head;
+         (nv != NULL);
+         nv = nv->l_next)
+    {
+        if (strcmp(nv->name, Name) == 0)
+        {
+            ret.append(nv->value);
+        }
+    }
+
+    return ret;
+}
+
+bool ExtUPnP::start()
+{
+    int error = 0;
+
+    /* The following devices are important */
+    static const char * const deviceList[] = {
+        "urn:schemas-upnp-org:device:MediaServer:1",
+        "urn:schemas-upnp-org:device:InternetGatewayDevice:1",
+        0
+    };
+    // TODO test other devices from upnp.org
+    /* more or less unimportant: */
+//        "urn:schemas-upnp-org:service:ConnectionManager:1",
+//        "urn:schemas-upnp-org:service:ContentDirectory:1",
+//        "urn:schemas-upnp-org:device:InternetGatewayDevice:2",
+//        "urn:schemas-upnp-org:service:WANIPConnection:2",
+//        "urn:schemas-upnp-org:service:WANIPConnection:1",
+//        "urn:schemas-upnp-org:service:WANPPPConnection:1",
+//        "upnp:rootdevice",
+//        "ssdp:all",
+
+    UPNPDev *devices = upnpDiscoverDevices(deviceList,
+                                           5000, NULL, NULL, FALSE,
+                                           FALSE, &error);
+    QList<ExtUPnPHash> list = goThroughDeviceList(devices);
+
+    /* Following functions can be called, see miniupnpc.h */
+//    UPNPDev *allDevList = upnpDiscoverAll(5000, NULL, NULL, FALSE, FALSE, &error);
+//    UPNPDev *IGDlist = ::upnpDiscover(2000, NULL, NULL, FALSE, FALSE, &error);
+//    UPNPDev *mediaServerList = ::upnpDiscoverDevice("urn:schemas-upnp-org:device:MediaServer:1",2000, NULL, NULL, FALSE, FALSE, &error);
+
+    emit finished();
+    return true; // TODO return false if something went wrong or if there are no results
+}
+
+bool ExtUPnP::stop()
+{
+    return true;
+}
+
+QList<ExtUPnP::ExtUPnPHash> ExtUPnP::goThroughDeviceList(UPNPDev *list)
+{
+    UPNPDev *listBegin = list;
+    QList<ExtUPnPHash> myResults;
+
+    while (list)
```

Comment:
Suggestion: use a for loop here and don't change `list`. The variable `listBegin` would then be obsolete, and your loop stuff would be together in one statement.
```cpp
for (UPNPDev *l = list; l; l = l->pNext)
...
int code = UPNP_GetValidIGD(l, &urls, &data, lanaddr, sizeof(lanaddr));
...
freeUPNPDevlist(list);
```